File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import *
from collections import deque
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        bfs: List[List[int]] = []
        if not root: return bfs
        queue = deque([root])

        while queue:
            size: int = len(queue)
            level: List[int] = [0] * size
            for i in range(size):
                node: Optional[TreeNode] = queue.popleft()
                level[i] = node.val

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            bfs.append(level)

        return bfs



        

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.ArrayList;
import java.util.List;

class Que {
    private ArrayList<TreeNode> queue;

    // Default constructor
    public Que(){
        this.queue = new ArrayList<>();
    }

    // Parameterized constructor
    public Que(ArrayList<TreeNode> queue){
        this.queue = queue;
    }

    public boolean isEmpty(){
        return this.queue.size() == 0;
    } 

    public TreeNode peak(){
        if(!this.isEmpty()){
            return this.queue.get(0);
        } else {
            return null; // Return null if the queue is empty
        }
    }

    public void enqueue(TreeNode node){
        this.queue.add(node);
    }

    public TreeNode dequeue(){
        if(!this.isEmpty()){
            return this.queue.remove(0);
        } else {
            return null; // Return null if the queue is empty
        }
    }
}

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {

        if(root == null){
            return new ArrayList<>();
        }

        List<List<Integer>> result = new ArrayList<>();

        Que queue = new Que();
        queue.enqueue(root);
        queue.enqueue(null);
        result.add(new ArrayList<>());

        while(!queue.isEmpty()){
            TreeNode currentNode = queue.dequeue();

            if (currentNode == null){
                if (!queue.isEmpty()){
                    result.add(new ArrayList<>());
                    queue.enqueue(null);
                }
            }else{
                result.get(result.size() - 1).add(currentNode.val);

                if(currentNode.left != null){
                    queue.enqueue(currentNode.left);
                }
                if(currentNode.right != null){
                    queue.enqueue(currentNode.right);
                }
                
            }
        }

        if (result.get(result.size()-1).size() == 0){
            result.remove(result.size()-1);
        }
        
        return result;
        
    }
}